### ISSUE TITLE
Grains terminate when fs.inotify.max_user_watches exceeded

### ISSUE THREAD

**Issue by jfrederickson** [sandstorm-io/sandstorm#3512]
I think I must have ended up with too many files in my grains, because I'd started to exceed the sysctl option `fs.inotify.max_user_watches` (which defaulted to 8192). This manifested as the grain I was working with terminating (with my nginx reverse proxy showing a 502 in the grain's iframe), and this showing up in the grain's debug logs:

```
2 3 4 5 6 7 8 | ** SANDSTORM SUPERVISOR: Starting up grain. Sandbox type: userns *** Uncaught exception *** sandstorm/supervisor.c++:232: overloaded: inotify_add_watch: No space left on device; path = sandbox/www_1611581607845905071/2010/06/the-open-order-of-lockpicks-toool-visits-hive stack: 420cf2 420775 420446 41f65d 415b66 414f04 41cc79 6da447 ** SANDSTORM SUPERVISOR: Starting up grain. Sandbox type: userns *** Uncaught exception *** sandstorm/supervisor.c++:232: overloaded: inotify_add_watch: No space left on device; path = sandbox/www_1611581607845905071/2010/06/summer-hackaton-2010 stack: 420cf2 420775 420446 41f65d 415b66 414f04 41cc79 6da447
-- | --
```

Increasing the value of the above sysctl key solved my problem. I'd been troubleshooting this in #sandstorm:

```
[7:19:06 pm] <isd> Yeah, that'd be the supervisor trying to keep track of how much storage a grain is using. Could you open an issue? Would be nice to handle this more gracefully.
```

**Comment by zenhack** [comment]
Perhaps we should just set this ourselves on startup if it is below a certain reasonable value?

**Comment by zenhack** [comment]
(It would be nice to find a more generally robust way to track space usage, but I don't know of one that wouldn't be major surgery. In general, I do not like the design of inotify).

**Comment by zenhack** [comment]
(Fwiw, the value on my laptop is 524288, having not changed it. On my live sandstorm box it is 14280. I don't know how these values are normally chosen).

**Comment by ocdtrekkie** [comment]
I did a bit of poking around, this topic has cropped up in various projects. This Stack Overflow felt valuable: https://stackoverflow.com/questions/535768/what-is-a-reasonable-amount-of-inotify-watches-with-linux

> I know from Lsyncd (maybe you want to check that out?) people who have a million watches. It just eats a Gigabyte of memory.

> A recent kernel change [1] dynamically selects a default max_user_watches value in the range [8192, 1048576] based on how much RAM the system has. (5.11 is the first kernel release containing this change.)

It seems like we probably could set this higher if it's too low, but it's definitely one of those system-specific settings which is why issues with it may be annoyingly frustrating to replicate at times. And whatever we set the value to, we are then just setting a new upper-bound on how many files a grain can have on a given box.

**Comment by zenhack** [comment]
Ok, well, hopefully as users upgrade to kernels that set this value based on available memory, this will be less and less of an issue. Available memory for the two systems I mentioned is 8GiB and 2GiB, respectively. We probably shouldn't override this then, since Sandstorm boxes are constrained by memory as-is.

It would be nice to find a way to monitor disk usage that didn't involve watching each file individually, but I don't know how to do this without storing each grain in its own file system (which is what blackrock did). Watching *every* file in *every* running grain feels very sad to me.

**Comment by ocdtrekkie** [comment]
Yeah, it seems like, especially considering the direct memory cost per file, a very frustrating constraint to scale.

**Comment by ocdtrekkie** [comment]
@Michael-S provided me the helpful instructions for this issue here (even though he has a different issue than this one, it seems):

> to inspect the value, do cat /proc/sys/fs/inotify/max_user_watches
> To change the value, you can change it dynamically but the easy way is to add a line to /etc/sysctl.conf:
> fs.inotify.max_user_watches=32768 and then restart.

I can endeavor to convert this tidbit into a docs section.

**Comment by zenhack** [comment]
Thinking about this again, and I discovered there's a comparatively new (kernel version 5.1) alternative to inotify; fanotify now supports watching a *mount point*, which would be less racy and generally more robust (and presumably use less memory than a watch per file). I haven't dug into the details, and I wouldn't be surprised if it was a little wonky wrt. bind mounts in one way or another, but something to think about.

**Comment by ocdtrekkie** [comment]
Would we retain the old behavior for older kernels, if this was a solution? I suspect we have Sandstorm servers out there running on the absolute minimum kernel we currently support.

**Comment by zenhack** [comment]
Yeah, we'd have to probably -- but it would at least make things better for most deployments.
